File: src/bot.py
"""
Contains code defining the bot and its behavior using Mineflayer and the llm module.
"""

from __future__ import annotations

import asyncio
from collections import namedtuple

from javascript import AsyncTask, On, Once, off, once, require

from llm import MinecraftCodeGenerator

mineflayer = require("mineflayer")
pathfinder = require("mineflayer-pathfinder")


def place_block(player, block_type, x, y, z):
    player.chat(f"/setblock {x} {y} {z} {block_type}")


Command = namedtuple("Command", ["handler", "description", "args"])

MESSAGE_COLOR = "light_purple"


class TextMCBot:
    def __init__(self, username: str = "TextMCBot", allowed_users: list[str] = None):
        self.username = username
        self.player = None
        self.allowed_users = allowed_users if allowed_users else []
        self.commands = None
        self.spawn_event = asyncio.Event()
        self.client = None
        self.movements = None

        self.connected = False

    async def connect(self, host: str, port: int) -> bool:
        connect_future = asyncio.Future()
        try:
            self.player = mineflayer.createBot(
                {
                    "host": host,
                    "port": port,
                    "username": self.username,
                    "hideErrors": True,  # Setting to false is good for debugging
                }
            )

            # Define handlers inline for clarity
            def handle_spawn(_):
                """Initial spawn handler just for connection verification"""
                if not connect_future.done():
                    connect_future.set_result(True)

            def handle_error(_, err):
                """Error handler for connection failures"""
                if not connect_future.done():
                    print(f"Connection error: {err}")
                    connect_future.set_result(False)

            # Register temporary handlers for connection
            self.player.once("spawn", handle_spawn)
            self.player.once("error", handle_error)

            # Wait for initial connection
            try:
                success = await asyncio.wait_for(connect_future, timeout=5.0)
                if not success:
                    return False

                # If connection successful, initialize bot
                self.player.loadPlugin(pathfinder.pathfinder)
                self.client = MinecraftCodeGenerator()
                self.movements = pathfinder.Movements(self.player)
                self.__bind_commands()
                self.__setup_listeners()
                self.__message(
                    "@a", "Hello! Chat $help for a list of commands", log=False
                )

                self.connected = True
                return True

            except asyncio.TimeoutError:
                print("Connection timed out")
                return False

        except Exception as e:
            print(f"Unexpected error during connection: {e}")
            return False

    def __message(self, recipient: str, message: str, log: bool = True):
        """Send a message to a recipient in-game

        :param recipient: The username of the recipient
        :type recipient: str
        :param message: The message to send
        ttype message: str
        """
        # Use the in-game /tellraw command to make the message look nice
        tellraw = f'/tellraw {recipient} {{"text":"[BOT] <{self.username}> {message}","color":"{MESSAGE_COLOR}"}}'
        if log:
            print(f"[BOT] <{self.username}> {message}")
        self.player.chat(tellraw)

    def __bind_commands(self):
        """Binds possible bot commands to their handlers"""
        self.commands = {
            "$come": Command(
                handler=self.command_come,
                description="Walk over to the sender.",
                args=[],
            ),
            "$build": Command(
                handler=self.command_build,
                description="Build a structure based on the sender's prompt.",
                args=["<prompt: string>"],
            ),
            "$where": Command(
                handler=self.command_where,
                description="Announce current location.",
                args=[],
            ),
            "$exec": Command(
                handler=self.command_exec,
                description="Run a Minecraft command (e.g., /tp, /setblock).",
                args=["<command: string>"],
            ),
            "$exit": Command(
                handler=self.command_exit,
                description="Disconnect from the world.",
                args=[],
            ),
            "$help": Command(
                handler=self.command_help,
                description="Show this help message.",
                args=[],
            ),
        }

    def __setup_listeners(self):
        @On(self.player, "chat")
        def on_chat(this, sender, message: str, *args):
            """Runs after every in-game chat message"""
            if not sender or sender == self.username:
                return
            if message.startswith("$"):  # Is a command?
                command, *args = message.split(" ")
                command = command.lower()
                command_obj = self.commands.get(command)
                if command_obj:
                    try:
                        command_obj.handler(sender, args)
                    except Exception as e:
                        print(f"Error executing command: {e}")
                        self.__message(
                            sender, "An error occurred while executing the command"
                        )
                else:
                    self.__message(
                        sender,
                        f"Unknown command: '{command}'. Try '$help' for a list of commands",
                    )

        @On(self.player, "end")
        def on_end(*args):
            """Called when the bot ends and disconnects from the server"""
            off(self.player, "chat", on_chat)
            off(self.player, "end", on_end)

    def command_help(self, sender, args):
        messages = ["Available commands:"]
        for command, cmd in self.commands.items():
            args_str = " ".join(cmd.args) if cmd.args else ""
            description = cmd.description if cmd.description else "No description"

            line = f"{command} {args_str} - {description}"
            messages.append(line)
        for msg in messages:
            self.__message(sender, msg)

    def command_where(self, sender, args):
        pos = self.player.entity.position
        self.__message(
            sender,
            f"I'm at X: {int(pos.x)}, Y: {int(pos.y)}, Z: {int(pos.z)} in the {self.player.game.dimension}",
        )

    def command_come(self, sender, args):
        player = self.player.players[sender]
        target = player.entity
        if not target:
            self.__message(sender, "I don't see you!")
            return
        pos = target.position
        self.player.pathfinder.setMovements(self.movements)
        try:
            self.player.pathfinder.setGoal(
                pathfinder.goals.GoalNear(pos.x, pos.y, pos.z, 1)
            )
        except Exception:
            self.__message(
                sender, "An error occurred with my pathfinding! Please try again."
            )

    def command_exec(self, sender, args):
        cmd = " ".join(args)
        if not cmd.startswith("/"):
            cmd = "/" + cmd
        self.player.chat(cmd)

    def command_build(self, sender, args):
        message = " ".join(args)  # Reconstructs the prompt
        response = self.client.generate_code(message)
        print("Generated code: ", response)
        try:
            self.execute_code(response)
        except RuntimeError as e:
            print("Error in generated code: ", e)
            self.__message(sender, "Error in generated code.")

    def command_exit(self, sender, args):
        self.connected = False
        self.__message("@a", "Bye! Disconnecting...")
        print("Bot disconnected")
        self.player.end()

    def execute_code(self, code):
        exec_context = {
            "place_block": place_block,
            "bot": self.player,
            "__builtins__": __builtins__,
        }
        try:
            # or... compile(code, "<string>", "exec") TODO: Consider this
            exec(code, exec_context)
        except Exception as e:
            raise RuntimeError(f"Failed to execute generated code: {e}")

File: src/bot/bot.ts
import { createBot, Bot } from "mineflayer";
import { pathfinder, Movements, goals } from "mineflayer-pathfinder";

import { parseCommand, getHelp } from "./commands";

const MESSAGE_COLOR = "light_purple";

/**
 * Create a mineflayer bot and connect it to a server
 * @param host The host of the server to connect to
 * @param port The port on that server
 * @param username The username to give to the bot's player
 * @param onDisconnect Optional callback for when the bot disconnects from the world
 * @returns Promise of the connected mineflayer bot
 */
export function createPlayer(
  host: string,
  port: number,
  username: string,
  onDisconnect?: (reason: string) => void
): Promise<Bot> {
  return new Promise((resolve, reject) => {
    const player = createBot({
      host: host,
      port: port,
      username: username,
      version: "1.20.4",
      hideErrors: false,
    });

    player.once("spawn", () => {
      player.removeAllListeners();
      resolve(player);
    });

    player.once("error", err => {
      reject(
        new Error(
          `Bot failed to connect to host '${host}' on port ${port}:\n${err}`
        )
      );
    });

    player.once("end", reason => {
      player.removeAllListeners();
      onDisconnect(reason);
    });
  });
}

// interface Command {
//   description: string;
//   handler: (sender: string, args: yargsParser.Arguments) => void;
// }

export class TextMCBot {
  private player: Bot;
  private exclusiveUsers: string[];
  private movements: Movements;

  /**
   * Take a mineflayer player and convert it to a TextMCBot, which encapsulates new features on top of mineflayer's capabilities
   *
   * @param player The mineflayer bot to convert to a TextMCBot
   * @param exclusiveUsers A list of usernames of players exclusively allowed to command the bot. If empty (default), anyone can command it
   */
  constructor(player: Bot, exclusiveUsers: string[] = []) {
    // I love dependency injection so much
    this.player = player;
    this.exclusiveUsers = exclusiveUsers;

    this.player.loadPlugin(pathfinder);
    this.movements = new Movements(this.player);

    this.message(
      "@a",
      `Hello! Send '${this.username}: help' in the chat for a list of commands.`
    );

    this.setupListeners();
  }

  get username() {
    return this.player.username;
  }

  /**
   * Configure mineflayer event listeners for the bot
   */
  private setupListeners() {
    // Fires when a chat message happens in the world
    this.player.on("chat", async (sender, msg) => {
      if (sender === this.username || !msg.startsWith(`${this.username}:`))
        return;

      if (
        this.exclusiveUsers.length > 0 &&
        !this.exclusiveUsers.includes(sender)
      ) {
        this.message(sender, "You are not in my list of exclusive users.");
        return;
      }

      // 'username: command arg1 arg2' -> 'command arg1 arg2'
      const cmd = msg.slice(this.username.length + 1);
      let parsed;
      try {
        parsed = parseCommand(cmd);
      } catch {
        this.message(
          sender,
          `Error reading command '${cmd}'. Try '${this.username}: help' for a list of commands.`
        );
        return;
      }

      // Switch by the command name (argument 0)
      switch (parsed._[0].toString()) {
        case "help": {
          const helpMsg = await getHelp();
          for (const line of helpMsg.split("\n")) {
            this.message(sender, line);
          }
          break;
        }
        case "exit": {
          this.player.end("exit command");
          break;
        }
        case "where": {
          const pos = this.player.entity.position;
          this.message(
            sender,
            `I'm at X: ${Math.floor(pos.x)}, Y: ${Math.floor(
              pos.y
            )}, Z: ${Math.floor(pos.z)} in the ${this.player.game.dimension}.`
          );
          break;
        }
        case "come": {
          const target = this.player.players[sender]?.entity;
          if (target) {
            const pos = target.position;
            this.player.pathfinder.setMovements(this.movements);
            // TODO: Consider a try-catch here just in case
            this.player.pathfinder.setGoal(
              new goals.GoalNear(pos.x, pos.y, pos.z, 1)
            );
          }
          break;
        }
        case "build": {
          // build handler
          break;
        }
      }
    });
  }

  /**
   * Send a message to a specific target
   *
   * @param target A username or a [Minecraft target selector](https://minecraft.wiki/w/Target_selectors) specifying who receives the message
   * @param msg The message to send
   */
  message(recipient: string, msg: string) {
    const tellraw = `/tellraw ${recipient} {"text":"[BOT] <${this.username}> ${msg}","color":"${MESSAGE_COLOR}"}`;
    this.player.chat(tellraw);
  }

  disconnect() {
    this.player.end("disconnect() function");
  }
}
